File: src/main/java/olivermakesco/de/refmagic/item/NecklaceItem.java
package olivermakesco.de.refmagic.item;

import dev.emi.trinkets.api.SlotReference;
import dev.emi.trinkets.api.TrinketItem;
import net.minecraft.client.item.TooltipContext;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.item.ItemStack;
import net.minecraft.item.PotionItem;
import net.minecraft.potion.PotionUtil;
import net.minecraft.text.Text;
import net.minecraft.text.TranslatableText;
import net.minecraft.util.Formatting;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.World;
import olivermakesco.de.refmagic.Mod;
import olivermakesco.de.refmagic.augment.Augment;
import olivermakesco.de.refmagic.augment.AugmentLoader;
import org.jetbrains.annotations.Nullable;
import org.lwjgl.system.CallbackI;

import java.util.AbstractMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class NecklaceItem extends TrinketItem {
    public NecklaceItem(Settings settings) {
        super(settings);
    }

    @Override
    public boolean isFireproof() {
        return true;
    }

    public static Map.Entry<Identifier, Augment> getAugment(ItemStack stack) {
        var nbt = stack.getOrCreateNbt();
        String name = "refmagic:base";
        if (nbt.contains("augment"))
            name = nbt.getString("augment");
        Identifier id = new Identifier((name));
        Augment augment = AugmentLoader.augments.get(id);
        if (augment == null)
            augment = AugmentLoader.augments.get(Mod.id("base"));
        return new AbstractMap.SimpleImmutableEntry<>(id, augment);
    }
    public static void setAugment(ItemStack stack, Identifier id) {
        var nbt = stack.getOrCreateNbt();
        nbt.putString("augment", id.toString());
        stack.setNbt(nbt);
    }

    @Override
    public ItemStack getDefaultStack() {
        var stack = super.getDefaultStack();
        setAugment(stack, Mod.id("base"));
        return stack;
    }

    int cooldown = 0;

    @Override
    public void tick(ItemStack stack, SlotReference slot, LivingEntity entity) {
        var augmentEntry = getAugment(stack);
        var augment = augmentEntry.getValue();
        cooldown++;
        cooldown %= 80;
        if (cooldown == 0)
            for (Identifier id : augment.potion())
                entity.addStatusEffect(new StatusEffectInstance(Registry.STATUS_EFFECT.get(id), 85));
    }

    @Override
    public String getTranslationKey(ItemStack stack) {
        Identifier id = Mod.id("base");
        try {
            var augmentEntry = getAugment(stack);
            id = augmentEntry.getKey();
        } catch (Throwable ignored) {}
        return id.getNamespace()+".necklace.augment."+id.getPath();
    }

    @Override
    public void appendTooltip(ItemStack stack, @Nullable World world, List<Text> tooltip, TooltipContext context) {
        var augments = getAugment(stack);
        for (Identifier potion : augments.getValue().potion())
            for (var instance : Registry.POTION.get(potion).getEffects())
                tooltip.add(
                        new TranslatableText(
                                "potion.withAmplifier",
                                new TranslatableText(instance.getTranslationKey()),
                                new TranslatableText("potion.potency."+instance.getAmplifier())
                        ).formatted(instance.getEffectType().getType().getFormatting())
                );
        super.appendTooltip(stack, world, tooltip, context);
    }
}

File: src/main/kotlin/olivermakesco/de/refmagic/item/NecklaceItem.kt
package olivermakesco.de.refmagic.item

import dev.emi.trinkets.api.SlotReference
import dev.emi.trinkets.api.TrinketItem
import net.minecraft.client.item.TooltipContext
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.effect.StatusEffectInstance
import net.minecraft.item.ItemStack
import net.minecraft.text.Text
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import net.minecraft.world.World
import olivermakesco.de.refmagic.Mod
import olivermakesco.de.refmagic.augment.Augment
import olivermakesco.de.refmagic.augment.AugmentLoader
import java.util.AbstractMap.SimpleImmutableEntry

class NecklaceItem(settings: Settings?) : TrinketItem(settings) {
    override fun isFireproof(): Boolean {
        return true
    }

    override fun getDefaultStack(): ItemStack {
        val stack = super.getDefaultStack()
        setAugment(stack, Mod.id("base"))
        return stack
    }

    var cooldown = 0
    override fun tick(stack: ItemStack, slot: SlotReference, entity: LivingEntity) {
        val (_, augment) = getAugment(stack)
        cooldown++
        cooldown %= 80
        if (cooldown == 0) for (id in augment.potion) entity.addStatusEffect(
            StatusEffectInstance(
                Registry.STATUS_EFFECT[id],
                85
            )
        )
    }

    override fun getTranslationKey(stack: ItemStack): String {
        var id = Mod.id("base")
        try {
            val (key) = getAugment(stack)
            id = key
        } catch (ignored: Throwable) {
        }
        return id.namespace + ".necklace.augment." + id.path
    }

    override fun appendTooltip(stack: ItemStack, world: World?, tooltip: MutableList<Text>, context: TooltipContext) {
        val (_, value) = getAugment(stack)
        for (potion in value.potion) for (instance in Registry.POTION[potion].effects) tooltip.add(
            Text.translatable(
                "potion.withAmplifier",
                Text.translatable(instance.translationKey),
                Text.translatable("potion.potency." + instance.amplifier)
            ).formatted(instance.effectType.type.formatting)
        )
        super.appendTooltip(stack, world, tooltip, context)
    }

    companion object {
        @JvmStatic
        fun getAugment(stack: ItemStack): Map.Entry<Identifier, Augment> {
            val nbt = stack.orCreateNbt
            var name: String? = "refmagic:base"
            if (nbt.contains("augment")) name = nbt.getString("augment")
            val id = Identifier(name)
            var augment = AugmentLoader.augments[id]
            if (augment == null) augment =
                Augment(0xffffff, ArrayList())
            return SimpleImmutableEntry(id, augment)
        }

        fun setAugment(stack: ItemStack, id: Identifier) {
            val nbt = stack.orCreateNbt
            nbt.putString("augment", id.toString())
            stack.nbt = nbt
        }
    }
}